crates/autopilot/src/domain/settlement/mod.rs

```diff
@@ -2,39 +2,151 @@
 ///
 /// Referenced as a [`Settlement`] in the codebase.
 #[allow(dead_code)]
+#[derive(Debug)]
 pub struct Settlement {
-    solution: Solution,
-    transaction: Transaction,
+    trades: Vec<Trade>,
+    auction: Auction,
 }
 
 impl Settlement {
-    pub async fn new(tx: eth::TxId, eth: &infra::Ethereum) -> Result<Self, Error> {
-        let transaction = eth.transaction(tx).await?;
-        let solution = Solution::new(
-            &transaction.input.0.clone().into(),
-            eth.contracts().settlement_domain_separator(),
-        )?;
-        Ok(Self {
-            solution,
-            transaction,
-        })
+    pub async fn new(
+        tx: &Transaction,
+        domain_separator: &eth::DomainSeparator,
+        persistence: &infra::Persistence,
+    ) -> Result<Self, Error> {
+        let tokenized::Tokenized {
+            tokens,
+            clearing_prices,
+            trades: decoded_trades,
+            interactions: _interactions,
+            auction_id,
+        } = tokenized::Tokenized::new(&tx.input)?;
+
+        let mut trades = Vec::with_capacity(decoded_trades.len());
+        for trade in decoded_trades {
+            let flags = tokenized::TradeFlags(trade.8);
+            let sell_token_index = trade.0.as_usize();
+            let buy_token_index = trade.1.as_usize();
+            let sell_token = tokens[sell_token_index];
+            let buy_token = tokens[buy_token_index];
+            let uniform_sell_token_index = tokens
+                .iter()
+                .position(|token| token == &sell_token)
+                .unwrap();
+            let uniform_buy_token_index =
+                tokens.iter().position(|token| token == &buy_token).unwrap();
+            trades.push(Trade::new(
+                tokenized::order_uid(&trade, &tokens, domain_separator)
+                    .map_err(|err| Error::OrderUidRecover(err, auction_id))?,
+                eth::Asset {
+                    token: sell_token.into(),
+                    amount: trade.3.into(),
+                },
+                eth::Asset {
+                    token: buy_token.into(),
+                    amount: trade.4.into(),
+                },
+                flags.side(),
+                trade.9.into(),
+                trade::Prices {
+                    uniform: trade::ClearingPrices {
+                        sell: clearing_prices[uniform_sell_token_index],
+                        buy: clearing_prices[uniform_buy_token_index],
+                    },
+                    custom: trade::ClearingPrices {
+                        sell: clearing_prices[sell_token_index],
+                        buy: clearing_prices[buy_token_index],
+                    },
+                },
+            ));
+        }
+
+        let auction = persistence.get_auction(auction_id).await?;
+        let solution = persistence.get_competition_solution(auction_id).await?;
+
+        let score = domain::competition::Score::new(
```

Comment:
nit: would move `Score` type on the top level domain, since it's no longer competition specific (it's just as much competition::Score as it would be settlement::Score)